components/script/dom/globalscope.rs

```diff
@@ -1254,23 +1253,64 @@ impl GlobalScope {
         &self,
         id: Uuid,
         promise: Rc<Promise>,
-        f: Box<dyn Fn(Rc<Promise>, Result<Vec<u8>, ()>) + Send>,
+        f: Box<dyn Fn(Rc<Promise>, Result<Vec<u8>, Error>) + Send>,
     ) {
         let recv = self.send_msg(id);
         let trusted_promise = TrustedPromise::new(promise);
         let canceller = self.task_canceller(TaskSourceName::FileReading);
         let task_source = self.file_reading_task_source();
-        thread::spawn(move || {
-            let bytes = GlobalScope::read_msg(recv);
-            let _ = task_source.queue_with_canceller(
-                task!(resolve_promise: move || {
-                    let promise = trusted_promise.root();
-                    let _ac = enter_realm(&*promise.global());
-                    f(promise, bytes);
-                }),
-                &canceller,
-            );
-        });
+
+        ROUTER.add_route(
+            recv.to_opaque(),
+            Box::new(move |message| {
+                let message = message.to();
+                let mut bytes = vec![];
```

Comment:
You also want to create the vec outside of the route, and move it into it, otherwise you will create a new one for each message received.

Reply:
I have made some progress here. First of all, I have used `bytes: DomRefCell<FileBytes>` in `FileListener`. But still I am some problems here.
1. I cannot assert `bytes` to be `Empty` or `Receiving` because `TrustedPromise` does not implement `Debug` or `PartialEq`. For now I have used `if let` to mutably borrows `bytes` out of `Receiving(bytes)` and in the `else` block used `panic!`.
2. I cannot figure out how to transform `Empty(trusted_promise)` to `Receiving(bytes, trusted_promise)`. I cannot transfer `trusted_promise` from one enum variant to another.  So I removed `TrustedPromise` from `FileBytes` in order to get past that.
3. Now I am facing a lifetime error. :disappointed: 

I have pushed the changes for you to review.

Reply:
I have removed the lifetime error with clone, but I cannot pass the closure(`f`) as parameter to `handle()` since it does not implement `Copy` or `Clone` trait. (same for `trusted_promise`, if passed as parameter instead of enum)